opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -426,309 +381,209 @@ namespace Opm {
-            BVector x(ebosJac.M());
             x = 0.0;
             linsolve.apply(x, ebosResid, result);
 
-            const int nw = wellModel().wells().number_of_wells;
-            BVector xw(nw);
+            // recover wells.
+            xw = 0.0;
             wellModel().recoverVariable(x, xw);
+        }
 
-            // convert to V;
+        //=====================================================================
+        // Implementation for ISTL-matrix based operator
+        //=====================================================================
 
+        /*!
+           \brief Adapter to turn a matrix into a linear operator.
 
-            V dx( (nc + nw) * np);
-            for( int p=0; p<np; ++p) {
-                for (int i = 0; i < nc; ++i) {
-                    int idx = i + nc*ebosCompToFlowPhaseIdx(p);
-                    dx(idx) = x[i][p];
-                }
-                for (int w = 0; w < nw; ++w) {
-                    int idx = w + nw*p + nc*np;
-                    dx(idx) = xw[w][flowPhaseToEbosCompIdx(p)];
-                }
-            }
+           Adapts a matrix to the assembled linear operator interface
+         */
+        template<class M, class X, class Y, class WellModel>
+        class WellModelMatrixAdapter : public Dune::MatrixAdapter<M,X,Y>
+        {
+          typedef Dune::MatrixAdapter<M,X,Y> BaseType;
+        public:
+          //! export types
+          typedef M matrix_type;
+          typedef X domain_type;
+          typedef Y range_type;
+          typedef typename X::field_type field_type;
+
+          //! constructor: just store a reference to a matrix
+          explicit WellModelMatrixAdapter (const M& A, WellModel& wellMod ) : BaseType( A ), wellMod_( wellMod ) {}
+
+          //! apply operator to x:  \f$ y = A(x) \f$
+          virtual void apply (const X& x, Y& y) const
+          {
+            BaseType::apply( x, y );
+
+            wellMod_.applyWellModel(x, y );
+          }
+
+        private:
+          WellModel& wellMod_;
+        };
 
-            //V dx2 = linsolver_.computeNewtonIncrement(residual_);
-            //std::cout << "------dx------- " << std::endl;
-            //std::cout << dx << std::endl;
-            //std::cout << "------dx2------- " << std::endl;
-            //std::cout << dx2 << std::endl;
+        /** @} end documentation */
 
-            //return dx;
-            return dx;
-        }
 
         /// Apply an update to the primary variables, chopped if appropriate.
         /// \param[in]      dx                updates to apply to primary variables
         /// \param[in, out] reservoir_state   reservoir state variables
         /// \param[in, out] well_state        well state variables
-        void updateState(const V& dx,
-                         ReservoirState& reservoir_state,
-                         WellState& well_state)
+        void updateState(const BVector& dx,
+                         ReservoirState& reservoir_state)
```

Comment:
This method updates the reservoir_state and is definitely needed. It does the Appelyard chopping and without it the solution would not converge. I implemented a much  simpler version of it when I refactored it. Moreover this new version improves the convergence significantly. Only tested on SPE1, SPE9 and Norne, so some more tuning may be needed.